File: 编译原理/TinyScript/src/main/java/parser/ast/Factor.java
package parser.ast;

import lexer.Token;
import lexer.TokenType;
import parser.util.PeekTokenIterator;

/**
 * 因子 - 操作符两边可以计算的内容
 */
public abstract class Factor extends ASTNode {
    public Factor(ASTNode _parent, PeekTokenIterator it) {
        super(_parent);
        Token token = it.next();
        TokenType type = token.getType();

        if (type == TokenType.VARIABLE) {
            this.type = ASTNodeTypes.VARIABLE;
        } else {
            this.type = ASTNodeTypes.SCALAR;
        }
        this.label = token.getValue();
        this.lexeme = token;
    }
}

File: 编译原理/TinyScript-Js/src/parser/ast/Factor.js
const ASTNode = require('./ASTNode')
const TokenType = require('../../lexer/TokenType')
const ASTNodeTypes = require('./ASTNodeTypes')

/**
 * 因子(抽象类) - 操作符两边可以计算的内容
 */
class Factor extends ASTNode {
  /**
   * 
   * @param {ASTNode} _parent
   * @param {PeekTokenIterator} it
   */
  constructor(_parent, it) {
    super(_parent)
    const token = it.next()
    const type = token.getType()

    if (type === TokenType.VARIABLE) {
      this.type = ASTNodeTypes.VARIABLE
    } else {
      this.type = ASTNodeTypes.SCALAR
    }

    this.label = token.getValue()
    this.lexeme = token
  }
}

module.exports = Factor